gutenberg.php

```diff
@@ -16,20 +16,203 @@
 defined( 'GUTENBERG_DEVELOPMENT_MODE' ) or define( 'GUTENBERG_DEVELOPMENT_MODE', true );
 ### END AUTO-GENERATED DEFINES
 
-gutenberg_pre_init();
+// Minimum supported version. Has to match "Requires at least" header above.
+if ( ! defined( 'GUTENBERG_MIN_WP_VERSION' ) ) {
+	define( 'GUTENBERG_MIN_WP_VERSION', '6.2' );
```

Comment:
Is the min version always known for each Gutenberg release?

What if the plan changes and thus that version might be different (maybe older such as a previous minor or later)? Later isn't necessarily problematic as the remedy is to upgrade. Older is problematic in that a fatal error might happen.

Wondering if instead, this minimum version should be set by Core and consumed in the plugin. For example, Core could provide a global function such as `_get_plugin_wp_min_compatible_version()` where for-core plugins like Gutenberg could invoke. Let Core determine which version is the min compatible.

Reply:
Uh, sorry but I think I'm missing something? Does that mean we're getting rid of the "Requires at least" plugin header?

> What if the plan changes and thus that version might be different

That means a plugin will not "know" what is the minimum WP version it supports at release, and want to change/fix it later? This hasn't been a problem... forever, not sure it is a problem now? 

> Wondering if instead, this minimum version should be set by Core...
> ...
> Let Core determine which version is the min compatible.

Not sure how core would "know" if a plugin is **fully compatible** or not. Emphasis on "fully compatible" here as some incompatibilities are not obvious or easy to detect.

Reply:
Whoops, this isn't correct. Sorry. It's a minimum version in Core, but a max version in the plugin.

>Not sure how core would "know" if a plugin is fully compatible or not. Emphasis on "fully compatible" here as some incompatibilities are not obvious or easy to detect.

Historically within Core itself, it views the minimum compatible version of the GB plugin to be the first version that causes a fatal error. That's very different from defining "fully compatible". It's focused on preventing fatal errors for users.

Reply:
Typically the versions that are compatible with the WordPress version in development aren't known until close to WP Beta 1. 

Just before Beta 1 or during beta is when the harded code GB version is updated Core (currently within `_upgrade_core_deactivate_incompatible_plugins()`).

What's my point?

The plugin's actual maximum WordPress version _might not_ be known at the plugin release time. It might change as the WordPress dev cycle advances into beta.

For example, a fatal error happened recently which then caused an update in Core to set the 17.6 as the minimum compat version for WP 6.5. I'm wondering if 17.5.x or older would have know it's max WP version is WP 6.4, and not 6.5?

Reply:
How can released GB plugin versions _know_ if Core itself will deactivate it due to be incompatible?

That's where the plugin could use the proposed Core function `_get_plugin_wp_min_compatible_version()`. 

Using the previous example, that Core function would return `17.6`. So then released plugin versions could invoke the Core function and take action itself. Then in time once all released versions of the plugin have this new code, Core would no longer need to deactivate the plugin, i.e. because the plugin handles it itself.

Reply:
> Whoops, this isn't correct. Sorry. It's a minimum version in Core, but a max version in the plugin.

Ahh I see. I knew I'm missing something :) 

> How can released GB plugin versions know if Core itself will deactivate it due to be incompatible?

The same way it "knows" it is not going to be installed on a server with a PHP version lower than the minimum required, and in a "too old" WP? I.e. the plugin sets its own requirements, and WP makes sure these requirements are met. Same as PHP and WP versions, and now plugin dependencies.

Similar logic has existed in core since 2007 (WP 2.0) I think: plugins can specify minimal WP version. This is opt-in and enforced by the plugins repo API and by the core plugins API. 

Adding another optional requirement for a maximum supported WP version would be best to follow the same logic/design imho, and should be available for all plugins. I know of 2-3 other plugins that can use this right now, for example the [Test jQuery Updates](https://wordpress.org/plugins/wp-jquery-update-test/) plugin that will probably be used again for jQuery 4.0.

Reply:
> The plugin's actual maximum WordPress version might not be known at the plugin release time. It might change as the WordPress dev cycle advances into beta.

Right. For that reason the "maximum supported" WP version can only be a major (two digits) version and should (generally) match the "tested up to" plugin header. 

Minor version of Gutenberg will still support the same "max WP" version. In addition the "max WP" version will include support for the next "WP Trunk", (alpha, beta, RC). This is needed in order to be able to use the current Gutenberg releases in a WP Trunk installs, like on wp.org. Please see the discussion above.